src/mainframe/job_cache.py

```diff
@@ -0,0 +1,224 @@
+import queue
+from datetime import datetime, timedelta, timezone
+from queue import Queue
+from typing import Optional
+
+import structlog
+from sqlalchemy import and_, or_, select, tuple_
+from sqlalchemy.orm import Session, selectinload
+
+from mainframe.constants import mainframe_settings
+from mainframe.models.orm import Rule, Scan, Status
+from mainframe.models.schemas import PackageScanResult, PackageScanResultFail
+
+logger: structlog.stdlib.BoundLogger = structlog.get_logger()
+
+
+class JobCache:
+    """Handles caching of jobs and results"""
+
+    def __init__(self, session: Session, size: int = 1) -> None:
+        self.scan_queue: Queue[Scan] = Queue(maxsize=size)
+        self.pending: list[Scan] = []
+        self.results_queue: Queue[PackageScanResult | PackageScanResultFail] = Queue(maxsize=size)
+        self.enabled = size > 1
```

Comment:
This works nicely for popping jobs off the queue, but not so much for pushing results onto the queue. If with a `maxsize` of 1, we can push a result onto the queue and it will not immediately get written to the database (which is the desired behaviour when caching is disabled).

One solution is to keep it the way I have, with a special `enabled` flag, another solution is to write all results to the DB after the last push. That is, if the cache size is 1, then we will first push a job onto the queue, check if it's full (yes, we just filled it up), and then write.

Reply:
Oh. Cache size 0 then. When getting, it will always fetch because it's always empty. When putting, it will always do IO because it's always full.

Reply:
Unfortunately setting the `maxsize` of a [`queue.Queue`](https://docs.python.org/3/library/queue.html) has the opposite effect:
>  If maxsize is less than or equal to zero, the queue size is infinite.

So while getting jobs will work fine (it's always empty), it will never be full and we'll never write the results to the DB